[dxva] Refactor the choice of conversions / restore support for RGB input

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
There was duplication between the three functions QueryHDRConversions, QueryHDRtoSDRConversions and QuerySDRConversions, now merged into one (their caller SupportedConversions()).
Most of the remaining code is actually for logging purposes.

There are no changes in the functionality for YCbCr pixel format (with the exception of commit 2 that fixes a small inefficiency without much impact added by the previous refactor)
Support for RGB sources was restored (temporary removed in previous refactor)

Logic of new SupportedConversions():
1. Using the provided source primaries, transfer, color space, range and destination type SDR or HDR
2. Override the transfer function for a few exceptions (already existing):
  - avoid tonemapping by dxva processor for HDR to SDR
  - apply custom shader transfer function conversion for HLG to HDR
  - rely on compatibility of HLG transfer function with SDR (primaries still converted by the processor)
3. Use the function AvToDxgiColorSpace to map:
  - the combination of AV primaries, transfer, color space and range for the source to a DXGI_COLOR_SPACE_TYPE value for processor input
  - the combination of output range + output SDR or HDR to a DXGI_COLOR_SPACE_TYPE value for processor output
4. Continue with previously existing logic to generate combinations of conversions for parameters above and test their support by the dxva processor.
5. If no supported conversions are found, try alternative chroma siting for HDR sources.

I believe the refactor work is mostly finished with this one and that the dxva renderer is now in pretty good shape.
After that it's functionality improvements and the work to avoid using the enumerator twice, once with GetWeight and once with Configure.

notes:
- RGB input support is done, but further changes outside of renderer needed to actually use it, see notes in Testing.
- I think it makes sense to create the RenderAsHDR() function in RendererBase because that's where the m_HdrType and HDR switch are decided (instead of having duplicate logic in CEnumeratorHD and RendererDXVA).
- github's diff viewer is doing a pretty bad job here, interlacing code of QueryHDRConverions with SupportedConversions, which don't have much in common.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Code duplication & maintainability.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 10, AMD 5600G, Intel 13th gen, Intel 4th gen, nVidia GT 1030
Tested with NV12 and P010 sources SDR, HDR and HLG
Output SDR, SDR on HDR capable, HDR. With and without "Switch HDR" enabled.
Full range/limited range screen.

Attempted to activate the RGB path, but after creating multiple samples, noticed that dxva decoder is disabled, and a filter graph is added automatically after software decode to convert everything to either NV12 or P010.
Additional changes would be needed to activate it, and probably at the same time make an improvement to avoid downgrading YUV 4:2:2 and 4:4:4 to 4:2:0, before chroma is converted back to 4:4:4 during rendering (with chroma resolution loss).

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
none

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [X] **None of the above** (please explain below)
Refactor

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[cleanup] unused constructor

CrystalP committed:
[dxva] Refactor the choice of conversions to check for support by the dxva processor

Reduce code duplication of QueryHDRConversions, QueryHDRtoSDRConversions and QuerySDRConversions.

Instead of special-casing the different situations, rely on AvToDxgiColorSpace to map the AV
attributes of the source to a dxgi color space.
Then apply a few exception rules
- do not use dxva tonemapping of HDR to SDR
- HLG custom conversion to HDR or SDR
- some dxva processors (nVidia) don't fully support topleft chroma siting > try alternative left siting.

This also brings back support for RGB input (not used in current code though because software decoded followed
by a conversion to NV12 or P010 before the renderer is reached)

afedchin commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.cpp:
```diff
@@ -766,3 +766,18 @@ DEBUG_INFO_VIDEO CRendererBase::GetDebugInfo(int idx)
   return info;
 }
+
+bool CRendererBase::RenderAsHDR(const AVColorPrimaries primaries,
```
```suggestion
bool CRendererBase::CanRenderAsHDR(const AVColorPrimaries primaries,
```

CrystalP replied:
With a different name if you don't mind (IntendToRenderAsHDR?). This is to communicate the intention of the rendering as SDR or HDR, because this will be called by GetWeight and Configure have to know before rendering actually starts, before refresh rate and/or Windows HDR switch.
It was not optimal to call the function from CheckVideoParameters, which should be fed with the actual mode of rendering as set by RendererBase (m_HdrType) > now done through new accessor ActualRenderAsHDR()
Although not well handled yet, it's possible the renderer may be unable to switch to the desired mode and there could be a difference between intention and actual. As a benefit the HDR switch enabled parameter is not needed anymore.

The related logic in RendererBase could be consolidated, but that wasn't the purpose of this PR, maybe later.

afedchin replied:
to be honest I don't remember all flow of the preparation of rendering now so I have no opinion at this. Just try

afedchin commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.cpp:
```diff
@@ -766,3 +766,18 @@ DEBUG_INFO_VIDEO CRendererBase::GetDebugInfo(int idx)
+
+  const bool canDisplayHDR = DX::Windowing()->IsHDROutput() || isHDRSwitchEnabled;
+
+  if (streamIsHDR && canDisplayHDR)
+    return true;
+  else
+    return false;
```
```suggestion
return streamIsHDR && canDisplayHDR;
```

CrystalP replied:
definitely!

afedchin reviewed (approved):
LGFM in general

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererDXVA.cpp:
```diff
@@ -33,15 +33,17 @@ void CRendererDXVA::GetWeight(std::map<RenderMethod, int>& weights, const VideoP
   {
-    if (enumerator.SupportedConversions(SupportedConversionsArgs{picture, systemUsesHDR}).empty())
+    const bool renderAsHDR =
+        RenderAsHDR(static_cast<AVColorPrimaries>(picture.color_primaries),
```
Seems this code block can be moved before  `if (av_pixel_format == AV_PIX_FMT_D3D11VA_VLD)`  to a avoid repeat the same code in `else` branch (only `RenderAsHDR` or even all block until  return `"DXVA will not be used."`)

CrystalP replied:
I split the original `if (enumerator.SupportedConversions())` lines for readability, the compiler should optimize that stuff away.
I suppose the reason there were 2 instead of just the one was for the few tests in the `av_pixel_format != AV_PIX_FMT_D3D11VA_VLD` case where flow doesn't get to `SupportedConversions`: `if (dxgi_format == DXGI_FORMAT_UNKNOWN)` and incompatible texture format.

I'll suggest a refactor in a separate commit.

CrystalP committed:
[dxva] Simplify GetWeight

The texture test could be integrated in SupportedConversions in a refactor.
Take Windows 8 vs Windows 10/11 path into account for proper logging.

CrystalP commented:
made changes above + runtime tested

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAEnumeratorHD.cpp:
```diff
@@ -546,238 +546,199 @@ ComPtr<ID3D11VideoProcessorOutputView> CEnumeratorHD::CreateVideoProcessorOutput
   }
-
-  // Try left chroma siting as fall back - half a pixel is not much difference at 4k
-  inputCS = DXGI_COLOR_SPACE_YCBCR_STUDIO_G22_LEFT_P2020;
```
`DXGI_COLOR_SPACE_YCBCR_STUDIO_G22_LEFT_P2020` this color space must be preserved because is used in all NVIDIA graphics in playback of regular 4K HDR10 sources to SDR display (tone mapped). Removing this causes render method DXVA is not usable for 4K HDR on these systems (all NVIDIA).

See: https://github.com/xbmc/xbmc/pull/23109#issuecomment-1500419723

thexai reviewed (changes requested):
NVIDIA compatibility should be addressed because is a big regression.

In general should be tested for both color spaces:

`DXGI_COLOR_SPACE_YCBCR_STUDIO_G22_LEFT_P2020`
`DXGI_COLOR_SPACE_YCBCR_STUDIO_G22_TOPLEFT_P2020`

It would also be desirable for HDR to HDR since this is working on master:

`DXGI_COLOR_SPACE_YCBCR_STUDIO_G2084_LEFT_P2020`
`DXGI_COLOR_SPACE_YCBCR_STUDIO_G2084_TOPLEFT_P2020`

Related commit: https://github.com/xbmc/xbmc/pull/23109/commits/5b528b9665e4e1f2d89d10756b38f038a63041f4

CrystalP commented:
OK the G22/P2020 needs work.

The G2084_TOPLEFT_P2020 is the standard supported by all as far as I know. But trying both alternatives may come for free because of the change for G22/P2020.

CrystalP commented:
New commit created for the G22/P2020 nVidia fix, will be merged into the others once approved.
I didn't find a very satisfying way to take care of G22/P2020 (at which level should the exception be added?). The proposed solution is open to discussion. From this solution it's trivial to use the file flags instead in the future.
The change does a bit more than is strictly necessary, but I think the extra bits are correct and will avoid WTFs in the future.
Tested with nVidia, AMD and Intel.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAEnumeratorHD.cpp:
```diff
@@ -546,238 +546,277 @@ ComPtr<ID3D11VideoProcessorOutputView> CEnumeratorHD::CreateVideoProcessorOutput
-                                                        AVColorPrimaries colorPrimaries,
-                                                        AVColorTransferCharacteristic colorTransfer)
+ProcessorConversions CEnumeratorHD::LogAndListConversions(
+    DXGI_FORMAT inputFormat,
```
Unused parameter `DXGI_FORMAT inputFormat` ?

CrystalP replied:
Absolutely. left over from copy/paste. I'll do one better, outputFormats was also not strictly needed, even if used.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererDXVA.cpp:
```diff
@@ -72,13 +54,26 @@ void CRendererDXVA::GetWeight(std::map<RenderMethod, int>& weights, const VideoP
+  const bool renderAsHDR =
+      IntendToRenderAsHDR(static_cast<AVColorPrimaries>(picture.color_primaries),
+                          static_cast<AVColorTransferCharacteristic>(picture.color_transfer));
+  if (enumerator.SupportedConversions({picture, renderAsHDR}).empty())
```
A blank line here will help improve readability

CrystalP replied:
It can be compacted yes, looked more readable with temp variable at some point but it's ok again. Changed.

About VideoPicture vs discrete parameters: it's a philosophical decision I guess and I could go either way.
When using only one or two members, I don't think it's worth using the VideoPicture, and what the function uses is clearer from the function signature. If it was 3 or more, then sure, too many parameters are less readable.

About the casts: a VideoPicture parameter moves the casts inside IntendToRender, it's not like they go away.
There could be a debate about defining better types inside VideoPicture: I understand it's not a great idea to use types/enums from an external library, but simply changing the type to unsigned int in VideoPicture and keeping the same meaning for each value is also not great.
I've been using the AV enums because they provide a bit more safety than simple int with no clear defined meaning, but I don't like it very much.

To do a better job, we'd probably need Kodi enums to represent the primaries, transfer, etc and conversion functions in one place.
Then we could avoid casts in many places, like here and a few others. Touches all platforms though.
What do you think?

thexai reviewed (approved):
The functional issues are solved, just two minor things.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAEnumeratorHD.cpp:
```diff
@@ -546,238 +546,277 @@ ComPtr<ID3D11VideoProcessorOutputView> CEnumeratorHD::CreateVideoProcessorOutput
+    default:
+      return AVCHROMA_LOC_UNSPECIFIED;
+  }
+}
```
Missing empty line at end

CrystalP replied:
yes

CrystalP commented:
Changes applied, further discussion possible about VideoPicture / discrete parameters or casts if you'd like.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererDXVA.cpp:
```diff
@@ -72,13 +54,28 @@ void CRendererDXVA::GetWeight(std::map<RenderMethod, int>& weights, const VideoP
+              {picture, IntendToRenderAsHDR(
+                            static_cast<AVColorPrimaries>(picture.color_primaries),
+                            static_cast<AVColorTransferCharacteristic>(picture.color_transfer))})
+          .empty())
```
Now it looks worse. If you don't want remove static_casts, leave it as before with the intermediate variable.

My idea was:

```c++
  if (enumerator.SupportedConversions({picture, IntendToRenderAsHDR(picture)}).empty())
  {
    CLog::LogF(LOGWARNING, "DXVA will not be used.");
    return;
  }
```

And in RendererBase:

```c++
bool CRendererBase::IntendToRenderAsHDR(const VideoPicture& picture)
{
  const bool streamIsHDR = (picture.color_primaries == AVCOL_PRI_BT2020) &&
                           (picture.color_transfer == AVCOL_TRC_SMPTE2084 ||
                            picture.color_transfer == AVCOL_TRC_ARIB_STD_B67);

  const bool canDisplayHDR =
      DX::Windowing()->IsHDROutput() || DX::Windowing()->IsHDRDisplaySettingEnabled();

  return streamIsHDR && canDisplayHDR;
}
```

static_cast are not necessary and has not cost pass `VideoPicture&` because is a reference (no object copy)

thexai replied:
Another way:

```c++
  const SupportedConversionsArgs args = {picture, IntendToRenderAsHDR(picture)};

  if (enumerator.SupportedConversions(args).empty())
  {
    CLog::LogF(LOGWARNING, "DXVA will not be used.");
    return;
  }
```

CrystalP replied:
ok I yield ;-) pushed with VideoPicture parameter.
If that's ok I'll squash most commits.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/DXVAEnumeratorHD.cpp:
```diff
@@ -546,238 +546,275 @@ ComPtr<ID3D11VideoProcessorOutputView> CEnumeratorHD::CreateVideoProcessorOutput
     {
-      for (const auto& c : convLeft)
+      for (const auto& c : conversions)
         CLog::LogF(LOGDEBUG, "supported conversion: {}", c.ToString());
```
```suggestion
        CLog::LogFC(LOGDEBUG, LOGVIDEO, "supported conversion: {}", c.ToString());
```

CrystalP replied:
good catch

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.h:
```diff
@@ -150,6 +150,22 @@ class CRendererBase
+   * \param transfer  transfer function of the source
+   * \return true: intend to render as HDR, false: intend to render as SDR
+   */
+  static bool IntendToRenderAsHDR(const VideoPicture& picture);
```
Update doxy parameters

CrystalP replied:
yes, done

thexai reviewed (approved):
👍

CrystalP commented:
Jenkins build this please

CrystalP commented:
Jenkins build this please
